File: solutions/Leetcode_48/leetcode_48.py
from typing import List


class Solution1:
    def rotate(self, matrix: List[List[int]]) -> None:
        n = len(matrix)
        for i in range(n//2):
            for j in range(n-n//2):
                matrix[i][j], matrix[~j][i], matrix[~i][~j], matrix[j][~i] = \
                    matrix[~j][i], matrix[~i][~j], matrix[j][~i], matrix[i][j]


class Solution2:
    def rotate(self, matrix):
        matrix[:] = zip(*matrix[::-1])

File: solutions/Leetcode_48/leetcode_48.cc
#include <algorithm>
#include <vector>

using namespace std;

class Solution1
{
public:
    void rotate(vector<vector<int>> &matrix)
    {
        int n = matrix.size();
        for (int i = 0; i < n / 2; ++i)
            for (int j = 0; j < n - n / 2; ++j)
            {
                swap(matrix[i][j], matrix[n - j - 1][i]);
                swap(matrix[n - j - 1][i], matrix[n - i - 1][n - j - 1]);
                swap(matrix[n - i - 1][n - j - 1], matrix[j][n - i - 1]);
            }
    }
};

class Solution2
{
public:
    void rotate(vector<vector<int>> &matrix)
    {
        reverse(matrix.begin(), matrix.end());
        for (int i = 0; i != matrix.size(); ++i)
            for (int j = i + 1; j != matrix.size(); ++j)
                swap(matrix[i][j], matrix[j][i]);
    }
};